File: server/src/services/fallbackService.ts
import { InventoryItem } from '../db';

export interface InsightResult {
  daysUntilEmpty: number;
  needsReorder: boolean;
  urgency: 'critical' | 'warning' | 'ok';
  reorderMessage: string;
  sustainabilityTip: string;
  alternativeSuppliers?: string[];
  source: 'ai' | 'fallback';
}

export function generateFallbackInsights(item: InventoryItem): InsightResult {
  const daysUntilEmpty = item.dailyUsageRate > 0
    ? Math.floor(item.quantity / item.dailyUsageRate)
    : 9999;

  let daysUntilExpiry: number | null = null;
  if (item.expiryDate) {
    const expiry = new Date(item.expiryDate);
    const now = new Date();
    daysUntilExpiry = Math.floor((expiry.getTime() - now.getTime()) / (1000 * 60 * 60 * 24));
  }

  const needsReorder = item.quantity <= item.reorderThreshold || daysUntilEmpty < 7;

  let urgency: 'critical' | 'warning' | 'ok';
  if (
    item.quantity <= 0 ||
    daysUntilEmpty <= 2 ||
    (daysUntilExpiry !== null && daysUntilExpiry <= 2)
  ) {
    urgency = 'critical';
  } else if (
    needsReorder ||
    daysUntilEmpty < 7 ||
    (daysUntilExpiry !== null && daysUntilExpiry <= 7)
  ) {
    urgency = 'warning';
  } else {
    urgency = 'ok';
  }

  let reorderMessage: string;
  if (urgency === 'critical') {
    if (item.quantity <= 0) {
      reorderMessage = `CRITICAL: ${item.name} is completely out of stock. Place an emergency order with ${item.supplier} immediately.`;
    } else if (daysUntilExpiry !== null && daysUntilExpiry <= 2) {
      reorderMessage = `CRITICAL: ${item.name} expires in ${daysUntilExpiry} day(s). Use or discard immediately and reorder from ${item.supplier}.`;
    } else {
      reorderMessage = `CRITICAL: ${item.name} will run out in approximately ${daysUntilEmpty} day(s). Order now from ${item.supplier}.`;
    }
  } else if (urgency === 'warning') {
    if (item.quantity <= item.reorderThreshold) {
      reorderMessage = `WARNING: ${item.name} is at or below reorder threshold (${item.quantity} ${item.unit} remaining, threshold: ${item.reorderThreshold}). Consider ordering from ${item.supplier}.`;
    } else if (daysUntilExpiry !== null && daysUntilExpiry <= 7) {
      reorderMessage = `WARNING: ${item.name} expires in ${daysUntilExpiry} day(s). Plan reorder from ${item.supplier} soon.`;
    } else {
      reorderMessage = `WARNING: ${item.name} will run out in approximately ${daysUntilEmpty} day(s). Plan a reorder from ${item.supplier}.`;
    }
  } else {
    reorderMessage = `${item.name} stock is adequate. Next reorder needed in approximately ${daysUntilEmpty} day(s). Current supplier: ${item.supplier}.`;
  }

  let sustainabilityTip: string;
  if (item.sustainabilityScore >= 9) {
    sustainabilityTip = `Excellent sustainability score (${item.sustainabilityScore}/10)! This item is a model eco-friendly choice. Share sourcing details with the team to encourage similar procurement decisions.`;
  } else if (item.sustainabilityScore >= 7) {
    sustainabilityTip = `Good sustainability score (${item.sustainabilityScore}/10). Consider asking ${item.supplier} about bulk purchasing discounts to reduce delivery emissions further.`;
  } else if (item.sustainabilityScore >= 5) {
    sustainabilityTip = `Moderate sustainability score (${item.sustainabilityScore}/10). Research eco-certified alternatives. Look for suppliers offering products with recyclable packaging or organic certifications.`;
  } else if (item.sustainabilityScore >= 3) {
    sustainabilityTip = `Low sustainability score (${item.sustainabilityScore}/10). Prioritize finding a greener supplier for ${item.name}. Check eco-procurement databases for certified alternatives.`;
  } else {
    sustainabilityTip = `Very low sustainability score (${item.sustainabilityScore}/10). This item should be replaced with an eco-friendly alternative as soon as possible. Consult your sustainability team for approved substitutes.`;
  }

  return {
    daysUntilEmpty,
    needsReorder,
    urgency,
    reorderMessage,
    sustainabilityTip,
    source: 'fallback',
  };
}

File: server/services/fallback_service.py
from datetime import datetime, timezone
from typing import Dict, Any, Optional


def _days_until_expiry(expiry_date: Optional[str]) -> float:
    if not expiry_date:
        return float("inf")
    try:
        expiry = datetime.fromisoformat(expiry_date.replace("Z", "+00:00"))
        now = datetime.now(timezone.utc)
        if expiry.tzinfo is None:
            expiry = expiry.replace(tzinfo=timezone.utc)
        return (expiry - now).days
    except Exception:
        return float("inf")


def generate_fallback_insights(item: Dict[str, Any]) -> Dict[str, Any]:
    quantity = item.get("quantity", 0)
    daily_rate = item.get("dailyUsageRate", 0)
    reorder_threshold = item.get("reorderThreshold", 0)
    expiry_date = item.get("expiryDate")
    sustainability_score = item.get("sustainabilityScore", 5)
    name = item.get("name", "item")
    unit = item.get("unit", "units")

    days_until_empty = quantity / daily_rate if daily_rate > 0 else float("inf")
    days_until_expiry = _days_until_expiry(expiry_date)

    needs_reorder = quantity <= reorder_threshold or days_until_empty <= 7

    if quantity <= reorder_threshold or days_until_empty <= 3 or days_until_expiry <= 2:
        urgency = "critical"
    elif days_until_empty <= 7 or days_until_expiry <= 7:
        urgency = "warning"
    else:
        urgency = "ok"

    if urgency == "critical":
        if days_until_expiry <= 2:
            reorder_message = f"CRITICAL: {name} expires very soon! Use or donate immediately to avoid waste."
        else:
            reorder_message = (
                f"CRITICAL: {name} is at or below reorder threshold "
                f"({reorder_threshold} {unit}). Reorder immediately."
            )
    elif urgency == "warning":
        days_str = f"{days_until_empty:.0f}" if days_until_empty != float("inf") else "unknown"
        reorder_message = f"Warning: {name} will run out in approximately {days_str} days. Consider reordering soon."
    else:
        days_str = f"{days_until_empty:.0f}" if days_until_empty != float("inf") else "plenty of"
        reorder_message = f"{name} is well-stocked. Estimated {days_str} days of supply remaining."

    if sustainability_score >= 9:
        tip = f"{name} already has an excellent sustainability score! Share your sourcing strategy with others."
    elif sustainability_score >= 7:
        tip = f"Good sustainability score for {name}. Consider switching to bulk purchasing to reduce packaging waste."
    elif sustainability_score >= 5:
        tip = f"Consider sourcing {name} from certified eco-friendly suppliers to improve your sustainability score."
    elif sustainability_score >= 3:
        tip = f"{name} has a low sustainability score. Look for recycled, refurbished, or locally-sourced alternatives."
    else:
        tip = (
            f"Priority improvement needed: {name} scores very low on sustainability. "
            "Research eco-certified alternatives immediately."
        )

    days_value = days_until_empty if days_until_empty != float("inf") else 9999

    return {
        "daysUntilEmpty": round(days_value, 1),
        "needsReorder": needs_reorder,
        "urgency": urgency,
        "reorderMessage": reorder_message,
        "sustainabilityTip": tip,
        "alternativeSuppliers": None,
        "source": "fallback",
    }
